Privacy assertion failure with `staged_api` and duplicated module name

Found with a [modified fuzz-rustc](https://github.com/jruderman/fuzz-rustc).

### Code

```Rust
#![feature(staged_api)]
pub mod a {
}
pub mod a {
    pub mod b {
    }
}
fn main(){}
```

<details><summary>Alternative input with stability attributes</summary>

```rust
#![feature(staged_api)]
#![stable(feature = "foo", since = "1.0")]

#[stable(feature = "foo", since = "1.0")]
pub mod a {
}
#[stable(feature = "foo", since = "1.0")]
pub mod a {
    #[stable(feature = "foo", since = "1.0")]
    pub mod b {
    }
}
fn main(){}
```

</details>

### Error output with debug assertions

```
error[E0428]: the name `a` is defined multiple times
error: internal compiler error: compiler/rustc_middle/src/middle/privacy.rs:171:17: private Restricted(DefId(0:4 ~ perr_n[96e5]::a#1)) > direct Restricted(DefId(0:5 ~ perr_n[96e5]::a#1::b))
```

<details><summary>Full output including backtrace</summary>

```
error[E0428]: the name `a` is defined multiple times
 --> perr-n.rs:4:1
  |
2 | pub mod a {
  | --------- previous definition of the module `a` here
3 | }
4 | pub mod a {
  | ^^^^^^^^^ `a` redefined here
  |
  = note: `a` must be defined only once in the type namespace of this module

error: internal compiler error: compiler/rustc_middle/src/middle/privacy.rs:171:17: private Restricted(DefId(0:4 ~ perr_n[96e5]::a#1)) > direct Restricted(DefId(0:5 ~ perr_n[96e5]::a#1::b))
 --> perr-n.rs:5:5
  |
5 |     pub mod b {
  |     ^^^^^^^^^

thread 'rustc' panicked at 'Box<dyn Any>', /Users/jruderman/code/rust/compiler/rustc_errors/src/lib.rs:967:33
stack backtrace:
   0:        0x105f67b51 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hca7435eb862ff36c
   1:        0x106004448 - core::fmt::write::hda6b0e10b55724fb
   2:        0x105f655c8 - std::io::Write::write_fmt::ha9d830225ddca362
   3:        0x105f67942 - std::sys_common::backtrace::print::h1790e63dbe903794
   4:        0x105f91355 - std::panicking::default_hook::{{closure}}::h43af9d0e518d235f
   5:        0x105f91137 - std::panicking::default_hook::h66fe7c7e4fb98f6f
   6:        0x117a99198 - rustc_driver[34272412877c8d7e]::DEFAULT_HOOK::{closure#0}::{closure#0}
   7:        0x105f91a16 - std::panicking::rust_panic_with_hook::h3f2771d8587a6225
   8:        0x11c6f9527 - std[8f57673cc279a434]::panicking::begin_panic::<rustc_errors[baeb4f2523d336bf]::ExplicitBug>::{closure#0}
   9:        0x11c6f9499 - std[8f57673cc279a434]::sys_common::backtrace::__rust_end_short_backtrace::<std[8f57673cc279a434]::panicking::begin_panic<rustc_errors[baeb4f2523d336bf]::ExplicitBug>::{closure#0}, !>
  10:        0x11ccf11a9 - std[8f57673cc279a434]::panicking::begin_panic::<rustc_errors[baeb4f2523d336bf]::ExplicitBug>
  11:        0x11c708769 - std[8f57673cc279a434]::panic::panic_any::<rustc_errors[baeb4f2523d336bf]::ExplicitBug>
  12:        0x11c7086e8 - <rustc_errors[baeb4f2523d336bf]::HandlerInner>::span_bug::<rustc_span[fb729f1b9e5ac969]::span_encoding::Span, &alloc[9c4954bf8315fe26]::string::String>
  13:        0x11c7084f4 - <rustc_errors[baeb4f2523d336bf]::Handler>::span_bug::<rustc_span[fb729f1b9e5ac969]::span_encoding::Span, &alloc[9c4954bf8315fe26]::string::String>
  14:        0x11c79c7d1 - rustc_middle[6f25fe7037d6027b]::ty::context::tls::with_opt::<rustc_middle[6f25fe7037d6027b]::util::bug::opt_span_bug_fmt<rustc_span[fb729f1b9e5ac969]::span_encoding::Span>::{closure#0}, ()>
  15:        0x11c79c841 - rustc_middle[6f25fe7037d6027b]::util::bug::opt_span_bug_fmt::<rustc_span[fb729f1b9e5ac969]::span_encoding::Span>
  16:        0x11cd0285f - rustc_middle[6f25fe7037d6027b]::util::bug::span_bug_fmt::<rustc_span[fb729f1b9e5ac969]::span_encoding::Span>
  17:        0x11c7b780b - <rustc_middle[6f25fe7037d6027b]::middle::privacy::EffectiveVisibilities>::check_invariants
  18:        0x11a5307d5 - rustc_privacy[d2e26e0ea95905a7]::effective_visibilities
  19:        0x11b7be7b0 - rustc_query_system[840d768d15d12085]::query::plumbing::try_execute_query::<rustc_query_impl[c109d582984dd181]::plumbing::QueryCtxt, rustc_query_system[840d768d15d12085]::query::caches::DefaultCache<(), &rustc_middle[6f25fe7037d6027b]::middle::privacy::EffectiveVisibilities>>
  20:        0x11b868c29 - rustc_query_system[840d768d15d12085]::query::plumbing::get_query::<rustc_query_impl[c109d582984dd181]::queries::effective_visibilities, rustc_query_impl[c109d582984dd181]::plumbing::QueryCtxt>
  21:        0x11b593027 - <rustc_query_impl[c109d582984dd181]::Queries as rustc_middle[6f25fe7037d6027b]::ty::query::QueryEngine>::effective_visibilities
  22:        0x11ad6fbe6 - rustc_passes[467a399e7fc15c73]::stability::check_unused_or_stable_features
  23:        0x117be16a7 - <rustc_session[c16dc8859e010203]::session::Session>::time::<(), rustc_interface[e9616f531c09701b]::passes::analysis::{closure#0}::{closure#2}::{closure#0}>
  24:        0x117bfa508 - std[8f57673cc279a434]::panicking::try::<(), core[63810368375d8c35]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[e9616f531c09701b]::passes::analysis::{closure#0}::{closure#2}>>
  25:        0x117be205c - <rustc_session[c16dc8859e010203]::session::Session>::time::<(), rustc_interface[e9616f531c09701b]::passes::analysis::{closure#0}>
  26:        0x117b9a311 - rustc_interface[e9616f531c09701b]::passes::analysis
  27:        0x11b7b61ed - rustc_query_system[840d768d15d12085]::query::plumbing::try_execute_query::<rustc_query_impl[c109d582984dd181]::plumbing::QueryCtxt, rustc_query_system[840d768d15d12085]::query::caches::DefaultCache<(), core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>>
  28:        0x11b8965d9 - rustc_query_system[840d768d15d12085]::query::plumbing::get_query::<rustc_query_impl[c109d582984dd181]::queries::analysis, rustc_query_impl[c109d582984dd181]::plumbing::QueryCtxt>
  29:        0x11b5647d7 - <rustc_query_impl[c109d582984dd181]::Queries as rustc_middle[6f25fe7037d6027b]::ty::query::QueryEngine>::analysis
  30:        0x117b173f8 - <rustc_interface[e9616f531c09701b]::passes::QueryContext>::enter::<rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>
  31:        0x117b01308 - <rustc_interface[e9616f531c09701b]::interface::Compiler>::enter::<rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}::{closure#2}, core[63810368375d8c35]::result::Result<core[63810368375d8c35]::option::Option<rustc_interface[e9616f531c09701b]::queries::Linker>, rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>
  32:        0x117a8729b - rustc_span[fb729f1b9e5ac969]::with_source_map::<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_interface[e9616f531c09701b]::interface::run_compiler<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}>::{closure#0}::{closure#1}>
  33:        0x117b0279a - <scoped_tls[a4e3500caacc884c]::ScopedKey<rustc_span[fb729f1b9e5ac969]::SessionGlobals>>::set::<rustc_interface[e9616f531c09701b]::interface::run_compiler<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}>::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>
  34:        0x117b065e0 - std[8f57673cc279a434]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e9616f531c09701b]::util::run_in_thread_pool_with_globals<rustc_interface[e9616f531c09701b]::interface::run_compiler<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}>::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>
  35:        0x117aea94d - std[8f57673cc279a434]::panicking::try::<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, core[63810368375d8c35]::panic::unwind_safe::AssertUnwindSafe<<std[8f57673cc279a434]::thread::Builder>::spawn_unchecked_<rustc_interface[e9616f531c09701b]::util::run_in_thread_pool_with_globals<rustc_interface[e9616f531c09701b]::interface::run_compiler<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}>::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  36:        0x117a91b80 - <<std[8f57673cc279a434]::thread::Builder>::spawn_unchecked_<rustc_interface[e9616f531c09701b]::util::run_in_thread_pool_with_globals<rustc_interface[e9616f531c09701b]::interface::run_compiler<core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>, rustc_driver[34272412877c8d7e]::run_compiler::{closure#1}>::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[63810368375d8c35]::result::Result<(), rustc_errors[baeb4f2523d336bf]::ErrorGuaranteed>>::{closure#1} as core[63810368375d8c35]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  37:        0x105f992eb - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3d6d78cd5f090dbf
  38:        0x105fa0eef - std::sys::unix::thread::Thread::new::thread_start::h54ce6bb2ca11d46f
  39:     0x7ff80d84d4e1 - __pthread_start

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.67.0-dev running on x86_64-apple-darwin

query stack during panic:
#0 [effective_visibilities] checking effective visibilities
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0428`.
```

</details>

### Regression

This is a new assertion, added in 448261a78a35026b3f5e855b705f35c916ecb19b of #103965 (@petrochenkov). I don't know whether this indicates a pre-existing bug.

### Version

Built stage1 on `x86_64-apple-darwin` at revision 01a6f30324deb8f9c9ec4a70c53690c5d073a244 with `debug-assertions = true`

--- a/compiler/rustc_middle/src/middle/privacy.rs
+++ b/compiler/rustc_middle/src/middle/privacy.rs
@@ -4,7 +4,6 @@
 use crate::ty::{DefIdTree, TyCtxt, Visibility};
 use rustc_data_structures::fx::FxHashMap;
 use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
-use rustc_hir::def::DefKind;
 use rustc_macros::HashStable;
 use rustc_query_system::ich::StableHashingContext;
 use rustc_span::def_id::LocalDefId;
@@ -185,7 +184,6 @@ impl EffectiveVisibilities {
                 );
             }
             let nominal_vis = tcx.visibility(def_id);
-            let def_kind = tcx.opt_def_kind(def_id);
             // FIXME: `rustc_privacy` is not yet updated for the new logic and can set
             // effective visibilities that are larger than the nominal one.
             if !nominal_vis.is_at_least(ev.reachable_through_impl_trait, tcx) && early {
@@ -196,11 +194,6 @@ impl EffectiveVisibilities {
                     ev.reachable_through_impl_trait,
                     nominal_vis
                 );
-            }
-            // Fully private items are never put into the table, this is important for performance.
-            // FIXME: Fully private `mod` items are currently put into the table.
-            if ev.reachable_through_impl_trait == private_vis && def_kind != Some(DefKind::Mod) {
-                span_bug!(span, "fully private item in the table {:?}: {:?}", def_id, ev.direct);
             }
         }
     }

--- a/compiler/rustc_middle/src/middle/privacy.rs
+++ b/compiler/rustc_middle/src/middle/privacy.rs
@@ -142,13 +142,13 @@ impl EffectiveVisibilities {
     pub fn set_public_at_level(
         &mut self,
         id: LocalDefId,
-        default_vis: impl FnOnce() -> Visibility,
+        lazy_private_vis: impl FnOnce() -> Visibility,
         level: Level,
     ) {
         let mut effective_vis = self
             .effective_vis(id)
             .copied()
-            .unwrap_or_else(|| EffectiveVisibility::from_vis(default_vis()));
+            .unwrap_or_else(|| EffectiveVisibility::from_vis(lazy_private_vis()));
         for l in Level::all_levels() {
             if l <= level {
                 *effective_vis.at_level_mut(l) = Visibility::Public;
@@ -206,6 +206,11 @@ impl EffectiveVisibilities {
     }
 }
 
+pub trait IntoDefIdTree {
+    type Tree: DefIdTree;
+    fn tree(self) -> Self::Tree;
+}
+
 impl<Id: Eq + Hash> EffectiveVisibilities<Id> {
     pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
         self.map.iter()
@@ -217,21 +222,26 @@ impl<Id: Eq + Hash> EffectiveVisibilities<Id> {
 
     // `parent_id` is not necessarily a parent in source code tree,
     // it is the node from which the maximum effective visibility is inherited.
-    pub fn update(
+    pub fn update<T: IntoDefIdTree>(
         &mut self,
         id: Id,
         nominal_vis: Visibility,
-        default_vis: Visibility,
+        lazy_private_vis: impl FnOnce(T) -> (Visibility, T),
         inherited_eff_vis: Option<EffectiveVisibility>,
         level: Level,
-        tree: impl DefIdTree,
+        mut into_tree: T,
     ) -> bool {
         let mut changed = false;
-        let mut current_effective_vis = self
-            .map
-            .get(&id)
-            .copied()
-            .unwrap_or_else(|| EffectiveVisibility::from_vis(default_vis));
+        let mut current_effective_vis = match self.map.get(&id).copied() {
+            Some(eff_vis) => eff_vis,
+            None => {
+                let private_vis;
+                (private_vis, into_tree) = lazy_private_vis(into_tree);
+                EffectiveVisibility::from_vis(private_vis)
+            }
+        };
+        let tree = into_tree.tree();
+
         if let Some(inherited_effective_vis) = inherited_eff_vis {
             let mut inherited_effective_vis_at_prev_level =
                 *inherited_effective_vis.at_level(level);

--- a/compiler/rustc_middle/src/middle/privacy.rs
+++ b/compiler/rustc_middle/src/middle/privacy.rs
@@ -213,14 +213,21 @@ impl<Id: Eq + Hash> EffectiveVisibilities<Id> {
         self.map.get(&id)
     }
 
-    // `parent_id` is not necessarily a parent in source code tree,
-    // it is the node from which the maximum effective visibility is inherited.
+    // FIXME: Share code with `fn update`.
+    pub fn effective_vis_or_private(
+        &mut self,
+        id: Id,
+        lazy_private_vis: impl FnOnce() -> Visibility,
+    ) -> &EffectiveVisibility {
+        self.map.entry(id).or_insert_with(|| EffectiveVisibility::from_vis(lazy_private_vis()))
+    }
+
     pub fn update<T: IntoDefIdTree>(
         &mut self,
         id: Id,
         nominal_vis: Visibility,
         lazy_private_vis: impl FnOnce(T) -> (Visibility, T),
-        inherited_eff_vis: Option<EffectiveVisibility>,
+        inherited_effective_vis: EffectiveVisibility,
         level: Level,
         mut into_tree: T,
     ) -> bool {
@@ -235,39 +242,36 @@ impl<Id: Eq + Hash> EffectiveVisibilities<Id> {
         };
         let tree = into_tree.tree();
 
-        if let Some(inherited_effective_vis) = inherited_eff_vis {
-            let mut inherited_effective_vis_at_prev_level =
-                *inherited_effective_vis.at_level(level);
-            let mut calculated_effective_vis = inherited_effective_vis_at_prev_level;
-            for l in Level::all_levels() {
-                if level >= l {
-                    let inherited_effective_vis_at_level = *inherited_effective_vis.at_level(l);
-                    let current_effective_vis_at_level = current_effective_vis.at_level_mut(l);
-                    // effective visibility for id shouldn't be recalculated if
-                    // inherited from parent_id effective visibility isn't changed at next level
-                    if !(inherited_effective_vis_at_prev_level == inherited_effective_vis_at_level
-                        && level != l)
-                    {
-                        calculated_effective_vis =
-                            if nominal_vis.is_at_least(inherited_effective_vis_at_level, tree) {
-                                inherited_effective_vis_at_level
-                            } else {
-                                nominal_vis
-                            };
-                    }
-                    // effective visibility can't be decreased at next update call for the
-                    // same id
-                    if *current_effective_vis_at_level != calculated_effective_vis
-                        && calculated_effective_vis
-                            .is_at_least(*current_effective_vis_at_level, tree)
-                    {
-                        changed = true;
-                        *current_effective_vis_at_level = calculated_effective_vis;
-                    }
-                    inherited_effective_vis_at_prev_level = inherited_effective_vis_at_level;
+        let mut inherited_effective_vis_at_prev_level = *inherited_effective_vis.at_level(level);
+        let mut calculated_effective_vis = inherited_effective_vis_at_prev_level;
+        for l in Level::all_levels() {
+            if level >= l {
+                let inherited_effective_vis_at_level = *inherited_effective_vis.at_level(l);
+                let current_effective_vis_at_level = current_effective_vis.at_level_mut(l);
+                // effective visibility for id shouldn't be recalculated if
+                // inherited from parent_id effective visibility isn't changed at next level
+                if !(inherited_effective_vis_at_prev_level == inherited_effective_vis_at_level
+                    && level != l)
+                {
+                    calculated_effective_vis =
+                        if nominal_vis.is_at_least(inherited_effective_vis_at_level, tree) {
+                            inherited_effective_vis_at_level
+                        } else {
+                            nominal_vis
+                        };
                 }
+                // effective visibility can't be decreased at next update call for the
+                // same id
+                if *current_effective_vis_at_level != calculated_effective_vis
+                    && calculated_effective_vis.is_at_least(*current_effective_vis_at_level, tree)
+                {
+                    changed = true;
+                    *current_effective_vis_at_level = calculated_effective_vis;
+                }
+                inherited_effective_vis_at_prev_level = inherited_effective_vis_at_level;
             }
         }
+
         self.map.insert(id, current_effective_vis);
         changed
     }

--- a/compiler/rustc_resolve/src/effective_visibilities.rs
+++ b/compiler/rustc_resolve/src/effective_visibilities.rs
@@ -155,32 +155,39 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         }
     }
 
-    fn effective_vis(&self, parent_id: ParentId<'a>) -> Option<EffectiveVisibility> {
-        match parent_id {
-            ParentId::Def(def_id) => self.def_effective_visibilities.effective_vis(def_id),
-            ParentId::Import(binding) => self.import_effective_visibilities.effective_vis(binding),
+    fn effective_vis_or_private(&mut self, parent_id: ParentId<'a>) -> EffectiveVisibility {
+        // Private nodes are only added to the table for caching, they could be added or removed at
+        // any moment without consequences, so we don't set `changed` to true when adding them.
+        *match parent_id {
+            ParentId::Def(def_id) => self
+                .def_effective_visibilities
+                .effective_vis_or_private(def_id, || self.r.private_vis_def(def_id)),
+            ParentId::Import(binding) => self
+                .import_effective_visibilities
+                .effective_vis_or_private(binding, || self.r.private_vis_import(binding)),
         }
-        .copied()
     }
 
     fn update_import(&mut self, binding: ImportId<'a>, parent_id: ParentId<'a>) {
         let nominal_vis = binding.vis.expect_local();
+        let inherited_eff_vis = self.effective_vis_or_private(parent_id);
         self.changed |= self.import_effective_visibilities.update(
             binding,
             nominal_vis,
             |r| (r.private_vis_import(binding), r),
-            self.effective_vis(parent_id),
+            inherited_eff_vis,
             parent_id.level(),
             &mut *self.r,
         );
     }
 
     fn update_def(&mut self, def_id: LocalDefId, nominal_vis: Visibility, parent_id: ParentId<'a>) {
+        let inherited_eff_vis = self.effective_vis_or_private(parent_id);
         self.changed |= self.def_effective_visibilities.update(
             def_id,
             nominal_vis,
             |r| (r.private_vis_def(def_id), r),
-            self.effective_vis(parent_id),
+            inherited_eff_vis,
             parent_id.level(),
             &mut *self.r,
         );

--- a/compiler/rustc_resolve/src/effective_visibilities.rs
+++ b/compiler/rustc_resolve/src/effective_visibilities.rs
@@ -10,6 +10,7 @@ use rustc_hir::def_id::CRATE_DEF_ID;
 use rustc_middle::middle::privacy::{EffectiveVisibilities, EffectiveVisibility};
 use rustc_middle::middle::privacy::{IntoDefIdTree, Level};
 use rustc_middle::ty::{DefIdTree, Visibility};
+use std::mem;
 
 type ImportId<'a> = Interned<'a, NameBinding<'a>>;
 
@@ -35,6 +36,8 @@ pub struct EffectiveVisibilitiesVisitor<'r, 'a> {
     /// keys in `Resolver::effective_visibilities` are not enough for that, because multiple
     /// bindings can correspond to a single def id in imports. So we keep a separate table.
     import_effective_visibilities: EffectiveVisibilities<ImportId<'a>>,
+    // It's possible to recalculate this at any point, but it's relatively expensive.
+    current_private_vis: Visibility,
     changed: bool,
 }
 
@@ -80,10 +83,12 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
             r,
             def_effective_visibilities: Default::default(),
             import_effective_visibilities: Default::default(),
+            current_private_vis: Visibility::Public,
             changed: false,
         };
 
         visitor.update(CRATE_DEF_ID, CRATE_DEF_ID);
+        visitor.current_private_vis = Visibility::Restricted(CRATE_DEF_ID);
         visitor.set_bindings_effective_visibilities(CRATE_DEF_ID);
 
         while visitor.changed {
@@ -155,6 +160,10 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         }
     }
 
+    fn cheap_private_vis(&self, parent_id: ParentId<'_>) -> Option<Visibility> {
+        matches!(parent_id, ParentId::Def(_)).then_some(self.current_private_vis)
+    }
+
     fn effective_vis_or_private(&mut self, parent_id: ParentId<'a>) -> EffectiveVisibility {
         // Private nodes are only added to the table for caching, they could be added or removed at
         // any moment without consequences, so we don't set `changed` to true when adding them.
@@ -170,11 +179,12 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
 
     fn update_import(&mut self, binding: ImportId<'a>, parent_id: ParentId<'a>) {
         let nominal_vis = binding.vis.expect_local();
+        let private_vis = self.cheap_private_vis(parent_id);
         let inherited_eff_vis = self.effective_vis_or_private(parent_id);
         self.changed |= self.import_effective_visibilities.update(
             binding,
             nominal_vis,
-            |r| (r.private_vis_import(binding), r),
+            |r| (private_vis.unwrap_or_else(|| r.private_vis_import(binding)), r),
             inherited_eff_vis,
             parent_id.level(),
             &mut *self.r,
@@ -182,11 +192,12 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
     }
 
     fn update_def(&mut self, def_id: LocalDefId, nominal_vis: Visibility, parent_id: ParentId<'a>) {
+        let private_vis = self.cheap_private_vis(parent_id);
         let inherited_eff_vis = self.effective_vis_or_private(parent_id);
         self.changed |= self.def_effective_visibilities.update(
             def_id,
             nominal_vis,
-            |r| (r.private_vis_def(def_id), r),
+            |r| (private_vis.unwrap_or_else(|| r.private_vis_def(def_id)), r),
             inherited_eff_vis,
             parent_id.level(),
             &mut *self.r,
@@ -213,8 +224,11 @@ impl<'r, 'ast> Visitor<'ast> for EffectiveVisibilitiesVisitor<'ast, 'r> {
             ),
 
             ast::ItemKind::Mod(..) => {
+                let prev_private_vis =
+                    mem::replace(&mut self.current_private_vis, Visibility::Restricted(def_id));
                 self.set_bindings_effective_visibilities(def_id);
                 visit::walk_item(self, item);
+                self.current_private_vis = prev_private_vis;
             }
 
             ast::ItemKind::Enum(EnumDef { ref variants }, _) => {

--- a/compiler/rustc_resolve/src/effective_visibilities.rs
+++ b/compiler/rustc_resolve/src/effective_visibilities.rs
@@ -42,6 +42,24 @@ impl Resolver<'_> {
     fn nearest_normal_mod(&mut self, def_id: LocalDefId) -> LocalDefId {
         self.get_nearest_non_block_module(def_id.to_def_id()).nearest_parent_mod().expect_local()
     }
+
+    fn private_vis_import(&mut self, binding: ImportId<'_>) -> Visibility {
+        let NameBindingKind::Import { import, .. } = binding.kind else { unreachable!() };
+        Visibility::Restricted(
+            import
+                .id()
+                .map(|id| self.nearest_normal_mod(self.local_def_id(id)))
+                .unwrap_or(CRATE_DEF_ID),
+        )
+    }
+
+    fn private_vis_def(&mut self, def_id: LocalDefId) -> Visibility {
+        if def_id == CRATE_DEF_ID {
+            Visibility::Public
+        } else {
+            Visibility::Restricted(self.nearest_normal_mod(def_id))
+        }
+    }
 }
 
 impl<'a, 'b> IntoDefIdTree for &'b mut Resolver<'a> {
@@ -143,36 +161,12 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         .copied()
     }
 
-    /// The update is guaranteed to not change the table and we can skip it.
-    fn is_noop_update(
-        &self,
-        parent_id: ParentId<'a>,
-        nominal_vis: Visibility,
-        default_vis: Visibility,
-    ) -> bool {
-        nominal_vis == default_vis
-            || match parent_id {
-                ParentId::Def(def_id) => self.r.visibilities[&def_id],
-                ParentId::Import(binding) => binding.vis.expect_local(),
-            } == default_vis
-    }
-
     fn update_import(&mut self, binding: ImportId<'a>, parent_id: ParentId<'a>) {
-        let NameBindingKind::Import { import, .. } = binding.kind else { unreachable!() };
         let nominal_vis = binding.vis.expect_local();
-        let default_vis = Visibility::Restricted(
-            import
-                .id()
-                .map(|id| self.r.nearest_normal_mod(self.r.local_def_id(id)))
-                .unwrap_or(CRATE_DEF_ID),
-        );
-        if self.is_noop_update(parent_id, nominal_vis, default_vis) {
-            return;
-        }
         self.changed |= self.import_effective_visibilities.update(
             binding,
             nominal_vis,
-            |r| (default_vis, r),
+            |r| (r.private_vis_import(binding), r),
             self.effective_vis(parent_id),
             parent_id.level(),
             &mut *self.r,
@@ -180,14 +174,10 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
     }
 
     fn update_def(&mut self, def_id: LocalDefId, nominal_vis: Visibility, parent_id: ParentId<'a>) {
-        let default_vis = Visibility::Restricted(self.r.nearest_normal_mod(def_id));
-        if self.is_noop_update(parent_id, nominal_vis, default_vis) {
-            return;
-        }
         self.changed |= self.def_effective_visibilities.update(
             def_id,
             nominal_vis,
-            |r| (if def_id == CRATE_DEF_ID { Visibility::Public } else { default_vis }, r),
+            |r| (r.private_vis_def(def_id), r),
             self.effective_vis(parent_id),
             parent_id.level(),
             &mut *self.r,

--- a/compiler/rustc_resolve/src/effective_visibilities.rs
+++ b/compiler/rustc_resolve/src/effective_visibilities.rs
@@ -7,7 +7,8 @@ use rustc_ast::EnumDef;
 use rustc_data_structures::intern::Interned;
 use rustc_hir::def_id::LocalDefId;
 use rustc_hir::def_id::CRATE_DEF_ID;
-use rustc_middle::middle::privacy::{EffectiveVisibilities, EffectiveVisibility, Level};
+use rustc_middle::middle::privacy::{EffectiveVisibilities, EffectiveVisibility};
+use rustc_middle::middle::privacy::{IntoDefIdTree, Level};
 use rustc_middle::ty::Visibility;
 
 type ImportId<'a> = Interned<'a, NameBinding<'a>>;
@@ -29,11 +30,25 @@ impl ParentId<'_> {
 
 pub struct EffectiveVisibilitiesVisitor<'r, 'a> {
     r: &'r mut Resolver<'a>,
+    def_effective_visibilities: EffectiveVisibilities,
     /// While walking import chains we need to track effective visibilities per-binding, and def id
     /// keys in `Resolver::effective_visibilities` are not enough for that, because multiple
     /// bindings can correspond to a single def id in imports. So we keep a separate table.
     import_effective_visibilities: EffectiveVisibilities<ImportId<'a>>,
     changed: bool,
+}
+
+impl Resolver<'_> {
+    fn nearest_normal_mod(&mut self, def_id: LocalDefId) -> LocalDefId {
+        self.get_nearest_non_block_module(def_id.to_def_id()).nearest_parent_mod().expect_local()
+    }
+}
+
+impl<'a, 'b> IntoDefIdTree for &'b mut Resolver<'a> {
+    type Tree = &'b Resolver<'a>;
+    fn tree(self) -> Self::Tree {
+        self
+    }
 }
 
 impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
@@ -43,6 +58,7 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
     pub fn compute_effective_visibilities<'c>(r: &'r mut Resolver<'a>, krate: &'c Crate) {
         let mut visitor = EffectiveVisibilitiesVisitor {
             r,
+            def_effective_visibilities: Default::default(),
             import_effective_visibilities: Default::default(),
             changed: false,
         };
@@ -54,6 +70,7 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
             visitor.changed = false;
             visit::walk_crate(&mut visitor, krate);
         }
+        visitor.r.effective_visibilities = visitor.def_effective_visibilities;
 
         // Update visibilities for import def ids. These are not used during the
         // `EffectiveVisibilitiesVisitor` pass, because we have more detailed binding-based
@@ -90,10 +107,6 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         info!("resolve::effective_visibilities: {:#?}", r.effective_visibilities);
     }
 
-    fn nearest_normal_mod(&mut self, def_id: LocalDefId) -> LocalDefId {
-        self.r.get_nearest_non_block_module(def_id.to_def_id()).nearest_parent_mod().expect_local()
-    }
-
     /// Update effective visibilities of bindings in the given module,
     /// including their whole reexport chains.
     fn set_bindings_effective_visibilities(&mut self, module_id: LocalDefId) {
@@ -124,7 +137,7 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
 
     fn effective_vis(&self, parent_id: ParentId<'a>) -> Option<EffectiveVisibility> {
         match parent_id {
-            ParentId::Def(def_id) => self.r.effective_visibilities.effective_vis(def_id),
+            ParentId::Def(def_id) => self.def_effective_visibilities.effective_vis(def_id),
             ParentId::Import(binding) => self.import_effective_visibilities.effective_vis(binding),
         }
         .copied()
@@ -150,7 +163,7 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         let default_vis = Visibility::Restricted(
             import
                 .id()
-                .map(|id| self.nearest_normal_mod(self.r.local_def_id(id)))
+                .map(|id| self.r.nearest_normal_mod(self.r.local_def_id(id)))
                 .unwrap_or(CRATE_DEF_ID),
         );
         if self.is_noop_update(parent_id, nominal_vis, default_vis) {
@@ -159,25 +172,25 @@ impl<'r, 'a> EffectiveVisibilitiesVisitor<'r, 'a> {
         self.changed |= self.import_effective_visibilities.update(
             binding,
             nominal_vis,
-            default_vis,
+            |r| (default_vis, r),
             self.effective_vis(parent_id),
             parent_id.level(),
-            ResolverTree(&self.r.definitions, &self.r.crate_loader),
+            &mut *self.r,
         );
     }
 
     fn update_def(&mut self, def_id: LocalDefId, nominal_vis: Visibility, parent_id: ParentId<'a>) {
-        let default_vis = Visibility::Restricted(self.nearest_normal_mod(def_id));
+        let default_vis = Visibility::Restricted(self.r.nearest_normal_mod(def_id));
         if self.is_noop_update(parent_id, nominal_vis, default_vis) {
             return;
         }
-        self.changed |= self.r.effective_visibilities.update(
+        self.changed |= self.def_effective_visibilities.update(
             def_id,
             nominal_vis,
-            if def_id == CRATE_DEF_ID { Visibility::Public } else { default_vis },
+            |r| (if def_id == CRATE_DEF_ID { Visibility::Public } else { default_vis }, r),
             self.effective_vis(parent_id),
             parent_id.level(),
-            ResolverTree(&self.r.definitions, &self.r.crate_loader),
+            &mut *self.r,
         );
     }
 

--- a/compiler/rustc_resolve/src/effective_visibilities.rs
+++ b/compiler/rustc_resolve/src/effective_visibilities.rs
@@ -9,7 +9,7 @@ use rustc_hir::def_id::LocalDefId;
 use rustc_hir::def_id::CRATE_DEF_ID;
 use rustc_middle::middle::privacy::{EffectiveVisibilities, EffectiveVisibility};
 use rustc_middle::middle::privacy::{IntoDefIdTree, Level};
-use rustc_middle::ty::Visibility;
+use rustc_middle::ty::{DefIdTree, Visibility};
 
 type ImportId<'a> = Interned<'a, NameBinding<'a>>;
 
@@ -54,10 +54,12 @@ impl Resolver<'_> {
     }
 
     fn private_vis_def(&mut self, def_id: LocalDefId) -> Visibility {
-        if def_id == CRATE_DEF_ID {
-            Visibility::Public
+        // For mod items `nearest_normal_mod` returns its argument, but we actually need its parent.
+        let normal_mod_id = self.nearest_normal_mod(def_id);
+        if normal_mod_id == def_id {
+            self.opt_local_parent(def_id).map_or(Visibility::Public, Visibility::Restricted)
         } else {
-            Visibility::Restricted(self.nearest_normal_mod(def_id))
+            Visibility::Restricted(normal_mod_id)
         }
     }
 }

--- a/src/test/ui/privacy/effective_visibilities.rs
+++ b/src/test/ui/privacy/effective_visibilities.rs
@@ -17,13 +17,13 @@ mod outer { //~ ERROR Direct: pub(crate), Reexported: pub(crate), Reachable: pub
         }
 
         #[rustc_effective_visibility]
-        struct PrivStruct; //~ ERROR not in the table
-                           //~| ERROR not in the table
+        struct PrivStruct; //~ ERROR Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
+                           //~| ERROR Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
 
         #[rustc_effective_visibility]
         pub union PubUnion { //~ ERROR Direct: pub(crate), Reexported: pub, Reachable: pub, ReachableThroughImplTrait: pub
             #[rustc_effective_visibility]
-            a: u8, //~ ERROR not in the table
+            a: u8, //~ ERROR Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
             #[rustc_effective_visibility]
             pub b: u8, //~ ERROR Direct: pub(crate), Reexported: pub, Reachable: pub, ReachableThroughImplTrait: pub
         }

--- a/src/test/ui/privacy/effective_visibilities.stderr
+++ b/src/test/ui/privacy/effective_visibilities.stderr
@@ -22,13 +22,13 @@ error: Direct: pub(crate), Reexported: pub, Reachable: pub, ReachableThroughImpl
 LL |         pub trait PubTrait {
    |         ^^^^^^^^^^^^^^^^^^
 
-error: not in the table
+error: Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
   --> $DIR/effective_visibilities.rs:20:9
    |
 LL |         struct PrivStruct;
    |         ^^^^^^^^^^^^^^^^^
 
-error: not in the table
+error: Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
   --> $DIR/effective_visibilities.rs:20:9
    |
 LL |         struct PrivStruct;
@@ -40,7 +40,7 @@ error: Direct: pub(crate), Reexported: pub, Reachable: pub, ReachableThroughImpl
 LL |         pub union PubUnion {
    |         ^^^^^^^^^^^^^^^^^^
 
-error: not in the table
+error: Direct: pub(self), Reexported: pub(self), Reachable: pub(self), ReachableThroughImplTrait: pub(self)
   --> $DIR/effective_visibilities.rs:26:13
    |
 LL |             a: u8,

--- a/src/test/ui/privacy/effective_visibilities_invariants.rs
+++ b/src/test/ui/privacy/effective_visibilities_invariants.rs
@@ -1,0 +1,12 @@
+// Invariant checking doesn't ICE in some cases with errors (issue #104249).
+
+#![feature(staged_api)] //~ ERROR module has missing stability attribute
+
+pub mod m {} //~ ERROR module has missing stability attribute
+
+pub mod m { //~ ERROR the name `m` is defined multiple times
+    // mod inner {} - ICE
+    type Inner = u8;
+}
+
+fn main() {}

--- a/src/test/ui/privacy/effective_visibilities_invariants.rs
+++ b/src/test/ui/privacy/effective_visibilities_invariants.rs
@@ -5,7 +5,7 @@
 pub mod m {} //~ ERROR module has missing stability attribute
 
 pub mod m { //~ ERROR the name `m` is defined multiple times
-    // mod inner {} - ICE
+    mod inner {}
     type Inner = u8;
 }
 

--- a/src/test/ui/privacy/effective_visibilities_invariants.stderr
+++ b/src/test/ui/privacy/effective_visibilities_invariants.stderr
@@ -1,0 +1,32 @@
+error[E0428]: the name `m` is defined multiple times
+  --> $DIR/effective_visibilities_invariants.rs:7:1
+   |
+LL | pub mod m {}
+   | --------- previous definition of the module `m` here
+LL |
+LL | pub mod m {
+   | ^^^^^^^^^ `m` redefined here
+   |
+   = note: `m` must be defined only once in the type namespace of this module
+
+error: module has missing stability attribute
+  --> $DIR/effective_visibilities_invariants.rs:3:1
+   |
+LL | / #![feature(staged_api)]
+LL | |
+LL | | pub mod m {}
+LL | |
+...  |
+LL | |
+LL | | fn main() {}
+   | |____________^
+
+error: module has missing stability attribute
+  --> $DIR/effective_visibilities_invariants.rs:5:1
+   |
+LL | pub mod m {}
+   | ^^^^^^^^^^^^
+
+error: aborting due to 3 previous errors
+
+For more information about this error, try `rustc --explain E0428`.

